content/intro-to-storybook/ember/en/data.md

```diff
@@ -0,0 +1,273 @@
+export const actions = {
+  ARCHIVE_TASK: 'ARCHIVE_TASK',
+  PIN_TASK: 'PIN_TASK',
+};
+
+// The action creators bundle actions with the data required to execute them
+export const archiveTask = id => ({ type: actions.ARCHIVE_TASK, id });
+export const pinTask = id => ({ type: actions.PIN_TASK, id });
+
+// All our reducers simply change the state of a single task.
+function taskStateReducer(taskState) {
+  return (state, action) => {
+    return {
+      ...state,
+      tasks: state.tasks.map(task =>
+        task.id === action.id ? { ...task, state: taskState } : task
+      ),
+    };
+  };
+}
+
+// The reducer describes how the contents of the store change for each action
+export const reducer = (state, action) => {
+  switch (action.type) {
+    case actions.ARCHIVE_TASK:
+      return taskStateReducer('TASK_ARCHIVED')(state, action);
+    case actions.PIN_TASK:
+      return taskStateReducer('TASK_PINNED')(state, action);
+    default:
+      return state || initialState;
+  }
+};
+
+export default combineReducers({
+  reducer,
+});
+```
+
+Then we'll update our `TaskList` to read data out of the store. First let's move our existing presentational version to `app/components/PureTaskList.hbs` and `app/components/PureTaskList.js` and wrap it with a container.
+
+In `app/components/PureTaskList.hbs`:
+
+```handlebars
+{{!--app/components/PureTaskList.hbs --}}
+
+
+{{!-- This file was moved from TaskList--}}
+{{#if (eq @loading true)}}
+  <LoadingRow />
+  <LoadingRow />
+  <LoadingRow/>
+  <LoadingRow />
+  <LoadingRow />
+{{/if}}
+
+{{#if this.emptyTasks}}
+  <div class="list-items">
+    <div class="wrapper-message">
+      <span class="icon-check" />
+      <div class="title-message">You have no tasks</div>
+      <div class="subtitle-message">Sit back and relax</div>
+    </div>
+  </div>
+{{/if}}
+
+{{#each this.tasksInOrder as |task|}}
+ {{Task task=task pinTask=(action "taskListOnPinTask") archiveTask=(action "taskListOnArchiveTask")}}
+{{/each}}
+```
+
+In `app/components/TaskList.hbs`:
+
+```handlebars
+
+{{!!-- app/components/TaskList.hbs --}}
+<div>
+   {{PureTaskList tasks=tasks pinTask=(action "taskListOnPinTask") archiveTask=(action "taskListOnArchiveTask")}}
+</div>
+
+```
+
+And `app/components/TaskList.js` :
+
+```js
+//app/components/TaskList.js
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { connect } from 'ember-redux';
+import { archiveTask, pinTask } from '../reducers/index';
+
+const stateToComputed = state => {
+  const { reducer } = state;
+  const { tasks } = reducer;
+  return {
+    tasks: tasks.filter(t => t.state === 'TASK_INBOX' || t.state === 'TASK_PINNED'),
+  };
+};
+
+class TaskList extends Component {
+  constructor() {
+    super(...arguments);
+  }
+
+  // the actions will bubble up to redux and update the store accordingly
+  @action
+  taskListOnPinTask(taskID) {
+    this.actions.onPinTask(taskID);
```

Comment:
I assume `this.actions` later comes in with `connect()`? And ember was finally soo proud to let the `actions` hash go 😂